File: Website/academy.cba.mit.edu/classes/input_devices/accel/hello.accel.45.c
//
// hello.accel.45.c
//
// accelerometer hello-world
//    9600 baud FTDI interface
//
// Neil Gershenfeld 11/1/14
// (c) Massachusetts Institute of Technology 2014
//
// This work may be reproduced, modified, distributed,
// performed, and displayed for any purpose. Copyright is
// retained and must be preserved. The work is provided
// as is; no warranty is provided, and users accept all 
// liability.
//

#include <avr/io.h>
#include <util/delay.h>

#define output(directions,pin) (directions |= pin) // set port direction for output
#define set(port,pin) (port |= pin) // set port pin
#define clear(port,pin) (port &= (~pin)) // clear port pin
#define pin_test(pins,pin) (pins & pin) // test for port pin
#define bit_test(byte,bit) (byte & (1 << bit)) // test for bit set
#define bit_delay_time 102 // bit delay for 9600 with overhead
#define bit_delay() _delay_us(bit_delay_time) // RS232 bit delay
#define half_bit_delay() _delay_us(bit_delay_time/2) // RS232 half bit delay

#define serial_port PORTB
#define serial_direction DDRB
#define serial_pin_out (1 << PB2)
#define in_port PORTB
#define in_direction DDRB
#define in_pins PINB
#define x_pin (1 << PB3)
#define y_pin (1 << PB4)

#define NLOOP 3

void put_char(volatile unsigned char *port, unsigned char pin, char txchar) {
   //
   // send character in txchar on port pin
   //    assumes line driver (inverts bits)
   //
   // start bit
   //
   clear(*port,pin);
   bit_delay();
   //
   // unrolled loop to write data bits
   //
   if bit_test(txchar,0)
      set(*port,pin);
   else
      clear(*port,pin);
   bit_delay();
   if bit_test(txchar,1)
      set(*port,pin);
   else
      clear(*port,pin);
   bit_delay();
   if bit_test(txchar,2)
      set(*port,pin);
   else
      clear(*port,pin);
   bit_delay();
   if bit_test(txchar,3)
      set(*port,pin);
   else
      clear(*port,pin);
   bit_delay();
   if bit_test(txchar,4)
      set(*port,pin);
   else
      clear(*port,pin);
   bit_delay();
   if bit_test(txchar,5)
      set(*port,pin);
   else
      clear(*port,pin);
   bit_delay();
   if bit_test(txchar,6)
      set(*port,pin);
   else
      clear(*port,pin);
   bit_delay();
   if bit_test(txchar,7)
      set(*port,pin);
   else
      clear(*port,pin);
   bit_delay();
   //
   // stop bit
   //
   set(*port,pin);
   bit_delay();
   //
   // char delay
   //
   bit_delay();
   }

int main(void) {
   //
   // main
   //
   static unsigned char i;
   static uint16_t xoff,xon,yoff,yon;
   //
   // set clock divider to /1 (8 MHz)
   //
   CLKPR = (1 << CLKPCE);
   CLKPR = (0 << CLKPS3) | (0 << CLKPS2) | (0 << CLKPS1) | (0 << CLKPS0);
   //
   // initialize output pins
   //
   set(serial_port, serial_pin_out);
   output(serial_direction, serial_pin_out);
   //
   // main loop
   //
   while (1) {
      //
      // read x (approximate loop timing)
      //
      xoff = 0;
      xon = 0;
      for (i = 0; i < NLOOP; ++i) {
         //
         // wait for x high 
         //
         while (0 == pin_test(in_pins,x_pin))
            ;
         //
         // wait for x low
         //
         while (0 != pin_test(in_pins,x_pin))
            ;
         //
         // time x low 
         //
         while (0 == pin_test(in_pins,x_pin))
            xoff += 1;
         //
         // time x high
         //
         while (0 != pin_test(in_pins,x_pin))
            xon += 1;
         }
      //
      // read y (approximate loop timing)
      //
      yoff = 0;
      yon = 0;
      for (i = 0; i < NLOOP; ++i) {
         //
         // wait for y high 
         //
         while (0 == pin_test(in_pins,y_pin))
            ;
         //
         // wait for y low
         //
         while (0 != pin_test(in_pins,y_pin))
            ;
         //
         // time y low 
         //
         while (0 == pin_test(in_pins,y_pin))
            yoff += 1;
         //
         // time y high
         //
         while (0 != pin_test(in_pins,y_pin))
            yon += 1;
         }
      //
      // send framing
      //
      put_char(&serial_port, serial_pin_out, 1);
      put_char(&serial_port, serial_pin_out, 2);
      put_char(&serial_port, serial_pin_out, 3);
      put_char(&serial_port, serial_pin_out, 4);
      //
      // send x
      //
      put_char(&serial_port, serial_pin_out, (xoff & 255));
      put_char(&serial_port, serial_pin_out, ((xoff >> 8) & 255));
      put_char(&serial_port, serial_pin_out, (xon & 255));
      put_char(&serial_port, serial_pin_out, ((xon >> 8) & 255));
      //
      // send y
      //
      put_char(&serial_port, serial_pin_out, (yoff & 255));
      put_char(&serial_port, serial_pin_out, ((yoff >> 8) & 255));
      put_char(&serial_port, serial_pin_out, (yon & 255));
      put_char(&serial_port, serial_pin_out, ((yon >> 8) & 255));
      }
   }

File: Website/academy.cba.mit.edu/classes/input_devices/accel/hello.accel.45.py
#
# hello.accel.45.py
#
# receive and display acceleration
#    hello.accel.45.py serial_port
#
# Neil Gershenfeld 11/3/14
# (c) Massachusetts Institute of Technology 2014
#
# This work may be reproduced, modified, distributed,
# performed, and displayed for any purpose. Copyright is
# retained and must be preserved. The work is provided
# as is; no warranty is provided, and users accept all 
# liability.
#

from Tkinter import *
import serial

WINDOW = 600 # window size

def idle(parent,canvas):
   global filter, eps
   #
   # idle routine
   #
   byte2 = 0
   byte3 = 0
   byte4 = 0
   ser.flush()
   while 1:
      #
      # find framing 
      #
      byte1 = byte2
      byte2 = byte3
      byte3 = byte4
      byte4 = ord(ser.read())
      if ((byte1 == 1) & (byte2 == 2) & (byte3 == 3) & (byte4 == 4)):
         break
   xoffl = ord(ser.read())
   xoffh= ord(ser.read())
   xonl = ord(ser.read())
   xonh= ord(ser.read())
   yoffl = ord(ser.read())
   yoffh= ord(ser.read())
   yonl = ord(ser.read())
   yonh= ord(ser.read())
   xoff = 256*xoffh+ xoffl
   xon = 256*xonh+ xonl
   xvalue = xoff/float(xon+xoff)
   x = int(.2*WINDOW + (.9-.2)*WINDOW*xvalue)
   canvas.itemconfigure("xtext",text="X %.2f"%xvalue)
   canvas.coords('xrect1',.2*WINDOW,.05*WINDOW,x,.2*WINDOW)
   canvas.coords('xrect2',x,.05*WINDOW,.9*WINDOW,.2*WINDOW)
   yoff = 256*yoffh+ yoffl
   yon = 256*yonh+ yonl
   yvalue = yoff/float(yon+yoff)
   y = int(.2*WINDOW + (.9-.2)*WINDOW*yvalue)
   canvas.itemconfigure("ytext",text="Y %.2f"%yvalue)
   canvas.coords('yrect1',.2*WINDOW,.3*WINDOW,y,.45*WINDOW)
   canvas.coords('yrect2',y,.3*WINDOW,.9*WINDOW,.45*WINDOW)
   canvas.update()
   parent.after_idle(idle,parent,canvas)

#
#  check command line arguments
#
if (len(sys.argv) != 2):
   print "command line: hello.accel.45.py serial_port"
   sys.exit()
port = sys.argv[1]
#
# open serial port
#
ser = serial.Serial(port,9600)
ser.setDTR()
#
# set up GUI
#
root = Tk()
root.title('hello.accel.45.py (q to exit)')
root.bind('q','exit')
canvas = Canvas(root, width=WINDOW, height=.5*WINDOW, background='white')
canvas.create_text(.1*WINDOW,.125*WINDOW,text=".33",font=("Helvetica", 24),tags="xtext",fill="#0000b0")
canvas.create_rectangle(.2*WINDOW,.05*WINDOW,.3*WINDOW,.2*WINDOW, tags='xrect1', fill='#b00000')
canvas.create_rectangle(.3*WINDOW,.05*WINDOW,.9*WINDOW,.2*WINDOW, tags='xrect2', fill='#0000b0')
canvas.create_text(.1*WINDOW,.375*WINDOW,text=".33",font=("Helvetica", 24),tags="ytext",fill="#0000b0")
canvas.create_rectangle(.2*WINDOW,.3*WINDOW,.3*WINDOW,.45*WINDOW, tags='yrect1', fill='#b00000')
canvas.create_rectangle(.3*WINDOW,.3*WINDOW,.9*WINDOW,.45*WINDOW, tags='yrect2', fill='#0000b0')
canvas.pack()
#
# start idle loop
#
root.after(100,idle,root,canvas)
root.mainloop()
